### PR TITLE
Improve origin array builder & make nullable/no-nullable builder support

### LINKED ISSUE #3692: Making `ArrayBuilder` write value to inner value directly, rather than call `arrow2`'s `try_push`
Now `ArrayBuilder` will call have `MutableArray` as the inner data structure. and then the inner `push` will add val with option, just like below:

https://github.com/datafuselabs/databend/blob/1b6bf394d7/common/datavalues/src/arrays/primitive/builder.rs#L37-L44

then the `push` will call arrow2's `try_push` to do real push

https://github.com/datafuse-extras/arrow2/blob/565cfaeb02/src/array/binary/mutable.rs#L407-L433

the code above will have extra branch predicting cost, so we will want to have write values dirctly to the values. and this will work  more effectively with no `Nullable` datatype, as mentioned in https://github.com/datafuselabs/databend/issues/3664

cc @sundy-li

### PR DESCRIPTION
**Pull request by Veeupup**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

* improve `MutableBooleanArrayBuilder`, `MutablePrimitiveArrayBuilder` and `MutableStringArrayBuilder`, to push value directly rather then call `try_push` to avoid branch prediction cost
* add nullable generic builder support. For example, if we determine that a certain column is with no nullable value, we can have a more efficient arraybuilder, just like below:

```rust
// with no nullable value
let builder = MutableStringArrayBuilder::<false>::default(); // false means no nullable, true means the value may be null
builder.push(val);
```

## Changelog

- Improvement

## Related Issues

Fixes #3692 

## Test Plan

Unit Tests

Stateless Tests

### CODE REVIEW

File: common/datavalues/src/arrays/boolean/mutable.rs [hunk 1]
```diff
@@ -71,15 +78,27 @@ impl MutableBooleanArrayBuilder {
...
             values: MutableBitmap::with_capacity(capacity),
-            validity: None,
+            validity: Some(MutableBitmap::with_capacity(capacity)),
         }
     }
 
-    pub fn from_data(values: MutableBitmap, validity: Option<MutableBitmap>) -> Self {
+    pub fn push(&mut self, value: bool) {
+        self.values.push(value);
+        self.validity.as_mut().unwrap().push(true);
```
Comment (sundy-li) [inline common/datavalues/src/arrays/boolean/mutable.rs:87]: Maybe no need to care about `validity` for `not nullable` case, the validity is always none.

File: common/datavalues/src/arrays/string/mutable.rs [hunk 2]
```diff
@@ -59,13 +59,20 @@ impl MutableArrayBuilder for MutableStringArrayBuilder {
...
 
-impl Default for MutableStringArrayBuilder {
+impl Default for MutableStringArrayBuilder<true> {
     fn default() -> Self {
         Self::new()
     }
 }
 
-impl MutableStringArrayBuilder {
+impl Default for MutableStringArrayBuilder<false> {
```
Comment (sundy-li) [inline common/datavalues/src/arrays/string/mutable.rs:68]: Can we combine them into one?

Comment (Veeupup) [reply common/datavalues/src/arrays/string/mutable.rs:68]: the compiler does not allow me to do so : (

File: common/datavalues/src/arrays/boolean/mutable.rs (hunk 1 above)
Comment (Veeupup) [reply common/datavalues/src/arrays/boolean/mutable.rs:87]: yes, it it redundant.

**Review by zhang2014 — APPROVED** [review]
Look good to me.

**Review by sundy-li — APPROVED** [review]

**Pull request merged**